File: C++/Pluszpont9/source.py
import sys
import math

class Csomopont:
    def __init__(self, b='/'):
        self.betu = b
        self.balNulla = None
        self.jobbEgy = None

class LZWBinFa:
    def __init__(self):
        self.gyoker = Csomopont('/')
        self.fa = self.gyoker
        self.melyseg = 0
        self.maxMelyseg = 0
        self.atlagosszeg = 0
        self.szorasosszeg = 0.0
        self.atlagdb = 0
        self.atlag = 0.0
        self.szoras = 0.0

    def push(self, b):
        if b == '0':
            if not self.fa.balNulla:
                self.fa.balNulla = Csomopont('0')
                self.fa = self.gyoker
            else:
                self.fa = self.fa.balNulla
        else:
            if not self.fa.jobbEgy:
                self.fa.jobbEgy = Csomopont('1')
                self.fa = self.gyoker
            else:
                self.fa = self.fa.jobbEgy

    def kiir(self, os):
        self.melyseg = 0
        self._kiir(self.gyoker, os)

    def _kiir(self, elem, os):
        if elem is not None:
            self.melyseg += 1
            self._kiir(elem.jobbEgy, os)
            for _ in range(self.melyseg):
                os.write("---")
            os.write(f"{elem.betu}({self.melyseg - 1})\n")
            self._kiir(elem.balNulla, os)
            self.melyseg -= 1

    def getMelyseg(self):
        self.melyseg = 0
        self.maxMelyseg = 0
        self.rmelyseg(self.gyoker)
        return self.maxMelyseg - 1

    def rmelyseg(self, elem):
        if elem is not None:
            self.melyseg += 1
            if self.melyseg > self.maxMelyseg:
                self.maxMelyseg = self.melyseg
            self.rmelyseg(elem.jobbEgy)
            self.rmelyseg(elem.balNulla)
            self.melyseg -= 1

    def getAtlag(self):
        self.melyseg = 0
        self.atlagosszeg = 0
        self.atlagdb = 0
        self.ratlag(self.gyoker)
        self.atlag = self.atlagosszeg / self.atlagdb if self.atlagdb > 0 else 0
        return self.atlag

    def ratlag(self, elem):
        if elem is not None:
            self.melyseg += 1
            self.ratlag(elem.jobbEgy)
            self.ratlag(elem.balNulla)
            self.melyseg -= 1
            if elem.jobbEgy is None and elem.balNulla is None:
                self.atlagdb += 1
                self.atlagosszeg += self.melyseg

    def getSzoras(self):
        self.atlag = self.getAtlag()
        self.szorasosszeg = 0.0
        self.melyseg = 0
        self.atlagdb = 0
        self.rszoras(self.gyoker)
        
        if self.atlagdb - 1 > 0:
            self.szoras = math.sqrt(self.szorasosszeg / (self.atlagdb - 1))
        else:
            self.szoras = math.sqrt(self.szorasosszeg)
        return self.szoras

    def rszoras(self, elem):
        if elem is not None:
            self.melyseg += 1
            self.rszoras(elem.jobbEgy)
            self.rszoras(elem.balNulla)
            self.melyseg -= 1
            if elem.jobbEgy is None and elem.balNulla is None:
                self.atlagdb += 1
                self.szorasosszeg += (self.melyseg - self.atlag) ** 2


def usage():
    print("Usage: lzwtree in_file -o out_file")

def main():
    if len(sys.argv) != 4:
        usage()
        sys.exit(-1)

    in_file = sys.argv[1]
    if sys.argv[2] != "-o":
        usage()
        sys.exit(-2)
    out_file = sys.argv[3]

    try:
        with open(in_file, "rb") as f_in:
            data = f_in.read()
    except IOError:
        print(f"{in_file} nem letezik...")
        usage()
        sys.exit(-3)

    # find first newline
    start_idx = 0
    while start_idx < len(data):
        if data[start_idx] == 0x0a:
            start_idx += 1
            break
        start_idx += 1

    binFa = LZWBinFa()
    kommentben = False

    for i in range(start_idx, len(data)):
        b = data[i]
        if b == 0x3e:  # '>'
            kommentben = True
            continue
        if b == 0x0a:  # '\n'
            kommentben = False
            continue
        if kommentben:
            continue
        if b == 0x4e:  # 'N'
            continue
        
        for j in range(8):
            if b & 0x80:
                binFa.push('1')
            else:
                binFa.push('0')
            b = (b << 1) & 0xff

    with open(out_file, "w") as f_out:
        binFa.kiir(f_out)
        f_out.write(f"depth = {binFa.getMelyseg()}\n")
        
        mean_val = binFa.getAtlag()
        mean_str = f"{mean_val:g}"
        if "." not in mean_str and "e" not in mean_str.lower() and mean_val != int(mean_val):
             mean_str = f"{mean_val:.5f}".rstrip("0").rstrip(".")

        var_val = binFa.getSzoras()
        
        f_out.write(f"mean = {mean_val:g}\n")
        f_out.write(f"var = {var_val:g}\n")


if __name__ == "__main__":
    main()

File: C++/Pluszpont9/source.ts
import * as fs from 'fs';

class Csomopont {
    betu: string;
    balNulla: Csomopont | null;
    jobbEgy: Csomopont | null;

    constructor(b: string = '/') {
        this.betu = b;
        this.balNulla = null;
        this.jobbEgy = null;
    }
}

class LZWBinFa {
    gyoker: Csomopont;
    fa: Csomopont;
    melyseg: number = 0;
    maxMelyseg: number = 0;
    atlagosszeg: number = 0;
    szorasosszeg: number = 0;
    atlagdb: number = 0;
    atlag: number = 0;
    szoras: number = 0;

    constructor() {
        this.gyoker = new Csomopont('/');
        this.fa = this.gyoker;
    }

    push(b: string) {
        if (b === '0') {
            if (!this.fa.balNulla) {
                this.fa.balNulla = new Csomopont('0');
                this.fa = this.gyoker;
            } else {
                this.fa = this.fa.balNulla;
            }
        } else {
            if (!this.fa.jobbEgy) {
                this.fa.jobbEgy = new Csomopont('1');
                this.fa = this.gyoker;
            } else {
                this.fa = this.fa.jobbEgy;
            }
        }
    }

    kiir(fd: number) {
        this.melyseg = 0;
        this._kiir(this.gyoker, fd);
    }

    private _kiir(elem: Csomopont | null, fd: number) {
        if (elem !== null) {
            this.melyseg++;
            this._kiir(elem.jobbEgy, fd);
            for (let i = 0; i < this.melyseg; i++) {
                fs.writeSync(fd, "---");
            }
            fs.writeSync(fd, `${elem.betu}(${this.melyseg - 1})\n`);
            this._kiir(elem.balNulla, fd);
            this.melyseg--;
        }
    }

    getMelyseg(): number {
        this.melyseg = 0;
        this.maxMelyseg = 0;
        this.rmelyseg(this.gyoker);
        return this.maxMelyseg - 1;
    }

    private rmelyseg(elem: Csomopont | null) {
        if (elem !== null) {
            this.melyseg++;
            if (this.melyseg > this.maxMelyseg) {
                this.maxMelyseg = this.melyseg;
            }
            this.rmelyseg(elem.jobbEgy);
            this.rmelyseg(elem.balNulla);
            this.melyseg--;
        }
    }

    getAtlag(): number {
        this.melyseg = 0;
        this.atlagosszeg = 0;
        this.atlagdb = 0;
        this.ratlag(this.gyoker);
        this.atlag = this.atlagdb > 0 ? this.atlagosszeg / this.atlagdb : 0;
        return this.atlag;
    }

    private ratlag(elem: Csomopont | null) {
        if (elem !== null) {
            this.melyseg++;
            this.ratlag(elem.jobbEgy);
            this.ratlag(elem.balNulla);
            this.melyseg--;
            if (elem.jobbEgy === null && elem.balNulla === null) {
                this.atlagdb++;
                this.atlagosszeg += this.melyseg;
            }
        }
    }

    getSzoras(): number {
        this.atlag = this.getAtlag();
        this.szorasosszeg = 0.0;
        this.melyseg = 0;
        this.atlagdb = 0;
        this.rszoras(this.gyoker);

        if (this.atlagdb - 1 > 0) {
            this.szoras = Math.sqrt(this.szorasosszeg / (this.atlagdb - 1));
        } else {
            this.szoras = Math.sqrt(this.szorasosszeg);
        }
        return this.szoras;
    }

    private rszoras(elem: Csomopont | null) {
        if (elem !== null) {
            this.melyseg++;
            this.rszoras(elem.jobbEgy);
            this.rszoras(elem.balNulla);
            this.melyseg--;
            if (elem.jobbEgy === null && elem.balNulla === null) {
                this.atlagdb++;
                this.szorasosszeg += Math.pow(this.melyseg - this.atlag, 2);
            }
        }
    }
}

function usage() {
        console.log("Usage: lzwtree in_file -o out_file");
}

function formatNumber(num: number): string {
    // Format to 6 significant digits to match C++'s default output (%g)
    return parseFloat(num.toPrecision(6)).toString();
}

function main() {
    const args = process.argv.slice(2);
    if (args.length !== 3) {
        usage();
        process.exit(-1);
    }

    const inFile = args[0];
    if (args[1] !== '-o') {
        usage();
        process.exit(-2);
    }
    const outFile = args[2];

    let data: Buffer;
    try {
        data = fs.readFileSync(inFile);
    } catch (err) {
        console.log(`${inFile} nem letezik...`);
        usage();
        process.exit(-3);
    }

    // Skip the first line as in original C++ impl loop condition
    let startIdx = 0;
    for (let i = 0; i < data.length; i++) {
        if (data[i] === 0x0a) {
            startIdx = i + 1;
            break;
        }
    }

    const binFa = new LZWBinFa();
    let kommentben = false;

    for (let i = startIdx; i < data.length; i++) {
        let b = data[i];

        if (b === 0x3e) { // '>'
            kommentben = true;
            continue;
        }
        if (b === 0x0a) { // '\n'
            kommentben = false;
            continue;
        }
        if (kommentben) continue;
        if (b === 0x4e) continue; // 'N'

        for (let j = 0; j < 8; j++) {
            if (b & 0x80) {
                binFa.push('1');
            } else {
                binFa.push('0');
            }
            b = (b << 1) & 0xff;
        }
    }

    const fd = fs.openSync(outFile, 'w');
    binFa.kiir(fd);
    fs.writeSync(fd, `depth = ${binFa.getMelyseg()}\n`);
    fs.writeSync(fd, `mean = ${formatNumber(binFa.getAtlag())}\n`);
    fs.writeSync(fd, `var = ${formatNumber(binFa.getSzoras())}\n`);
    fs.closeSync(fd);
}

main();